cartographer_ros/cartographer_ros/assets_writer.cc

```diff
@@ -236,47 +256,30 @@ void RunPipeline(
-                             const std::string& configuration_directory,
-                             const std::string& configuration_basename,
-                             const std::string& urdf_filename,
-                             const std::string& output_file_prefix,
-                             const bool use_bag_transforms) {
-  carto::mapping::proto::PoseGraph pose_graph_proto;
-  carto::mapping::proto::AllTrajectoryBuilderOptions
-      all_trajectory_builder_options;
-  std::tie(pose_graph_proto, all_trajectory_builder_options) =
+void AssetsWriter::Init(const std::string& pose_graph_filename,
+                        const std::string& configuration_directory,
+                        const std::string& configuration_basename,
+                        const std::string& output_file_prefix) {
+  std::tie(pose_graph_, trajectory_options_) =
       LoadPoseGraph(pose_graph_filename);
-
-  CHECK_EQ(pose_graph_proto.trajectory_size(), bag_filenames.size())
-      << "Pose graphs contains " << pose_graph_proto.trajectory_size()
-      << " trajectories while " << bag_filenames.size()
+  CHECK_EQ(pose_graph_->trajectory_size(), bag_filenames_.size())
+      << "Pose graphs contains " << pose_graph_->trajectory_size()
+      << " trajectories while " << bag_filenames_.size()
       << " bags were provided. This tool requires one bag for each "
          "trajectory in the same order as the correponding trajectories in the "
          "pose graph proto.";
-
   // This vector must outlive the pipeline.
-  std::vector<carto::mapping::proto::Trajectory> all_trajectories(
-      pose_graph_proto.trajectory().begin(),
-      pose_graph_proto.trajectory().end());
-  const std::string& file_prefix = !output_file_prefix.empty()
-                                       ? output_file_prefix
-                                       : bag_filenames.front() + "_";
-  auto pipeline_builder = CreatePipelineBuilder(all_trajectories, file_prefix);
-
-  auto lua_parameter_dictionary =
-      CreateLuaDictionary(configuration_directory, configuration_basename);
-  std::vector<std::unique_ptr<carto::io::PointsProcessor>> pipeline =
-      pipeline_builder->CreatePipeline(
-          lua_parameter_dictionary->GetDictionary("pipeline").get());
-  const std::string& tracking_frame =
-      lua_parameter_dictionary->GetString("tracking_frame");
+  all_trajectories_.resize(pose_graph_->trajectory_size());
+  std::copy(pose_graph_->trajectory().begin(), pose_graph_->trajectory().end(),
+            all_trajectories_.begin());
 
-  RunPipeline(pipeline, bag_filenames, pose_graph_proto, tracking_frame,
-              urdf_filename, use_bag_transforms);
+  const std::string file_prefix = !output_file_prefix.empty()
+                                      ? output_file_prefix
+                                      : bag_filenames_.front() + "_";
+  point_pipeline_builder_ =
+      CreatePipelineBuilder(all_trajectories_, file_prefix);
+  lua_parameter_dictionary_ =
```

Comment:
Why not move this next to `CreatePipeline()` like before?

Reply:
In the next PR the class will allow to register custom points-processors. That will require that the creation of the pipeline is postponed until Run() is called.

Reply:
Not sure I understand your answer. When you move loading the Lua parameter dictionary to `Run()`, the pipeline is still built there, right?